src/parser/rogue/shared/shadowlands/covenants/necrolord/SerratedBoneSpike.tsx

```diff
@@ -0,0 +1,76 @@
+import React from 'react';
+import Analyzer, { SELECTED_PLAYER } from 'parser/core/Analyzer';
+import Abilities from 'parser/core/modules/Abilities';
+import SpellUsable from 'parser/shared/modules/SpellUsable';
+import SPELLS from 'common/SPELLS';
+import Events, { DamageEvent, EnergizeEvent, RemoveDebuffEvent } from 'parser/core/Events';
+import RESOURCE_TYPES from 'game/RESOURCE_TYPES';
+import Statistic from 'interface/statistics/Statistic';
+import STATISTIC_CATEGORY from 'interface/others/STATISTIC_CATEGORY';
+import COVENANTS from 'game/shadowlands/COVENANTS';
+import ItemDamageDone from 'interface/ItemDamageDone';
+import BoringSpellValueText from 'interface/statistics/components/BoringSpellValueText';
+import ResourceIcon from 'common/ResourceIcon';
+
+class SerratedBoneSpike extends Analyzer {
+  static dependencies = {
+    abilities: Abilities,
+    spellUsable: SpellUsable,
+  };
+  damage: number = 0;
+  comboPointsGained: number = 0;
+  comboPointsWasted: number = 0;
+
+  protected abilities!: Abilities;
+  protected spellUsable!: SpellUsable;
+
+  constructor(options: any) {
+    super(options);
+    this.active = this.selectedCombatant.hasCovenant(COVENANTS.NECROLORD.id);
+    this.addEventListener(Events.damage.by(SELECTED_PLAYER).spell(SPELLS.SERRATED_BONE_SPIKE), this.onDamage);
+    this.addEventListener(Events.energize.by(SELECTED_PLAYER).spell(SPELLS.SERRATED_BONE_SPIKE_ENERGIZE), this.onEnergize);
+    this.addEventListener(Events.removedebuff.by(SELECTED_PLAYER).spell(SPELLS.SERRATED_BONE_SPIKE_DEBUFF), this.onSBSRemoveDebuff);
+  }
+
+  onDamage(event: DamageEvent) {
+    this.damage += event.amount + (event.absorbed || 0);
+  }
+
+  onEnergize(event: EnergizeEvent) {
+    if (event.resourceChangeType === RESOURCE_TYPES.COMBO_POINTS.id) {
+      this.comboPointsGained += event.resourceChange;
+      this.comboPointsWasted += event.waste;
+    }
+  }
+
+  onSBSRemoveDebuff(event: RemoveDebuffEvent) {
```

Comment:
Im not sure what the spell does ... But if it is just straight CDR, you dont need to do the math regarding the remaining CD. Just reduce it by however long the CDR from the spell is ... if its more than the remaining CD of the ability, it will just end the CD automatically.

Reply:
When an enemy dies while affected by Serrated Bone Spike a charge is refunded. This needs to effectively end the cooldown and smset the remaining cooldown for the next charge to be equal to the cooldown remaining of the ability before the charge was refunded for an enemy dying.

Reply:
Then wouldnt it be simpler to just reduce the Cooldown by the recharge time of the spell?

So like Fire Blast has 3 charges and a 10 sec recharge. So if i had used all 3 charges, and there was 4 seconds left until I gained a charge ... if i reduced the cooldown by 10 seconds then i would have a charge available and 4 seconds until i gained another.

Reply:
If the recharge timer is affected by haste (like Fire Blast) you would need to include that in the reduction as well.

this.spellUsable.reduceCooldown(SPELLS.SPELL.id, 10 / (1 + haste))

Reply:
I wasn't aware it had multiple charges - you can see Lock and Load for utilizing endCooldown with multiple charges (it lets the overflow CDR go towards next charge)

Reply:
So if I just endCooldown that will automagically reduce the remaining cooldown for the next charge(if applicable)?

Reply:
no.

You reduce the cooldown by the full recharge time.

So if it takes 10 seconds to gain a charge (i.e. the spell has a 10 sec recharge), then reduceCooldown by 10 seconds.

Serrated Bone Spike has a 30 sec recharge, so reduce the cooldown by 30 seconds if you want to gain 1 full charge while retaining the time until the next charge.

Reply:
if the reduction passed to reduceCooldown, is larger than the CD it actually calls endCooldown in this manner. 

```js
if (cooldownRemaining < reductionMs) {
      const remainingCDR = reductionMs - cooldownRemaining;
      return cooldownRemaining + this.endCooldown(canSpellId, false, timestamp, remainingCDR);
```

Reply:
ahh, i still find it easier and simpler to just do the reduce without having to do all the math on what is left on the CD and such. but thats me

Reply:
So should I keep what I have or change it?

Reply:
`reduceCooldown` is the easier call to use in this case. see [here](https://github.com/WoWAnalyzer/WoWAnalyzer/blob/264a7c07a92075071ffffec5cca4e4c9a488cd55/src/parser/paladin/protection/modules/core/GrandCrusader.js#L61-L62) for how judgment refunds are done for prot paladin (the ability has 2 charges with the talent that gives refunds).